plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/AggregationQueryPageSource.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.elasticsearch;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.plugin.elasticsearch.client.ElasticsearchClient;
+import io.trino.plugin.elasticsearch.client.composite.CompositeAggregation;
+import io.trino.plugin.elasticsearch.decoders.Decoder;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.connector.ConnectorPageSource;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.aggregations.Aggregation;
+import org.elasticsearch.search.aggregations.Aggregations;
+import org.elasticsearch.search.aggregations.metrics.NumericMetricsAggregation;
+import org.elasticsearch.search.aggregations.metrics.stats.Stats;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.elasticsearch.ElasticsearchQueryBuilder.buildAggregationQuery;
+import static io.trino.plugin.elasticsearch.ElasticsearchQueryBuilder.buildSearchQuery;
+import static java.util.Objects.requireNonNull;
+
+public class AggregationQueryPageSource
+        implements ConnectorPageSource
+{
+    private long readTimeNanos;
+    private final List<Decoder> decoders;
+    private final BlockBuilder[] columnBuilders;
+    private final List<ElasticsearchColumnHandle> columns;
+    private final ElasticsearchClient client;
+    private final ElasticsearchTableHandle table;
+    private final ElasticsearchSplit split;
+    private final int pageSize;
+    // the after parameter that composite aggregation used to fetch by page
+    private Optional<Map<String, Object>> after;
+    private boolean fetched;
+    private static final SearchHit noUsed = new SearchHit(0);
```

Comment:
@martint Since the Decoder defined `SearchHit` as parameter, we have to pass SearchHit instance to `decode`. But when push aggregation down, we never need `SearchHit`. So `noUsed` just act as a placeholder.

Reply:
@martint Since the Decoder defined `SearchHit` as parameter, we have to pass SearchHit instance to `decode`. But when push aggregation down, we never need `SearchHit`. So `noUsed` just act as a placeholder.